### DISCUSSION TITLE
ucore：关于ucore用户态程序中包含字符串常量指针的变量未能正确初始化的问题

### DISCUSSION THREAD

**Discussion by lklake** [LearningOS/rust-based-os-comp2022#181] [General]
# 问题
ucore ch5练习中，在usershell运行ch5_mergetest，会产生如下错误：
![image](https://user-images.githubusercontent.com/36527804/233787261-d60a7af7-39ea-4cea-a672-df2667765693.png)

ch5_mergetest代码如下：

![image](https://user-images.githubusercontent.com/36527804/233787412-38c1b81f-1252-458a-afae-4832cf725178.png)

可以看到ch5_mergetest并未得到待测试程序的名称，因而打印出的目标测试程序名称均为`null`。

通过测试发现数组TESTS和FAILES中的数据均为0，于是ch5_mergetest获得的程序名均为空指针，导致错误。

进一步实验发现，ucore用户态程序中包含常量字符串指针的变量，都无法正常初始化（其初始值始终为0），包括
1. `char *[]`类型的全局常量字符串指针数组。
2. 或者`char *`类型的全局常量字符串指针。
3. 另外，当栈上`char*[]`类型的常量字符串指针数组大小达到一定值时，也会无法初始化。
```c
// 1
char *TESTS[] = {"waku"};
int main(){
    printf("%s\n",TESTS[0]);
}
```
```c
// 2
char *TESTS = "waku";
int main(){
    printf("%s\n",TESTS);
}
```
```c
// 3
int main(){
    char *TESTS[] = {"waku",some more};
    printf("%s\n",TESTS[0]);
}
```
# 环境
ucore实验指导中的工具链与qemu。
- riscv64-unknown-elf-gcc-10.1.0-2020.08.2-x86_64-linux-ubuntu14
- riscv64-linux-musl-cross，gcc version = 10.2.1 binutils version = 2.36.1
- qemu-7.0.0
# 复现
修改Makefile，将ch5_mergetest添加到ch5的测试程序集（ch5原始Makefile好像并未包括该程序），并在usershell中执行ch5_mergetest；
或者在任一用户程序中创建全局变量`char * str = "waku";`，并在main函数中尝试打印。
# 原因推测
链接器在链接ucore用户态程序时，会把TESTS和FAILS放到名为.data.rel.local的section中，初始化为0，且该段在链接器链接过程中中符号没有重定位。导致最终TESTS、FAILS等数组内容为空。（猜测可能是由于TESTS、FAILS等全局变量引用了常量字符串，被编译器优化为在 **运行时** 进行重定位，而ucore不提供运行时重定位功能，导致问题发生。
![image](https://user-images.githubusercontent.com/36527804/233787112-2223a1c9-d6fa-434b-a4e3-0fc85748a502.png)
# 解决办法
在用户程序的链接选项中添加`-static`参数，可以避免这一问题。

![image](https://user-images.githubusercontent.com/36527804/233788776-ed41f701-b09c-4002-a9cd-d3c79bcd6268.png)

# 其他
通过测试发现：
- 如果是全局变量包含常量字符串地址，该全局变量一定会出现该问题；
- 如果是局部数组型变量包含常量字符串的地址，当数组长度到一定程度时，也会出现该问题。。推测可能是由于编译器的某种优化策略导致。

## 欢迎大家交流讨论，比如也遇到过这个问题或者了解gcc相关文档。😁

**Comment by haozixu** [comment]
也遇到过这个问题，我的解决方案是给用户程序编译选项加上-no-pie和-fno-pie。当时也有点纳闷为什么生成的是ELF shared object而不是executable。我猜有可能现在链接中的musl gcc是默认开启生成PIE的，而原来的那个不是。

**Reply by lklake** [reply]
确实唉。原项目生成的用户态elf格式可执行程序是ELF shared object，有dynamic segment：
![image](https://user-images.githubusercontent.com/36527804/233827529-b547cbe9-ea11-48b7-9077-a85b457b26bd.png)

加上-static后就是elf executable了，没有dynamic segment：

![image](https://user-images.githubusercontent.com/36527804/233827691-5cd17dcd-4623-4ecf-be2a-a3a5b4e4598a.png)

不过现在常量字符串都尝试使用动态链接来优化了？挺厉害，不过提到这个的文档不太好找。
